packages/spring/src/index.ts

```diff
@@ -0,0 +1,215 @@
+export type SpringSetter<T> = (
+  newValue: T | ((prev: T) => T),
+  opts?: SpringSetterOptions,
+) => Promise<void>;
+
+/**
+ * Creates a signal and a setter that uses spring physics when interpolating from
+ * one value to another. This means when the value changes, instead of
+ * transitioning at a steady rate, it "bounces" like a spring would,
+ * depending on the physics paramters provided. This adds a level of realism to
+ * the transitions and can enhance the user experience.
+ *
+ * `T` - The type of the signal. It works for the basic data types that can be
+ * interpolated: `number`, a `Date`, `Array<T>` or a nested object of T.
+ *
+ * @param initialValue The initial value of the signal.
+ * @param options Options to configure the physics of the spring.
+ * @returns Returns the spring value and a setter.
+ *
+ * @example
+ * const [progress, setProgress] = createSpring(0, { stiffness: 0.15, damping: 0.8 });
+ */
+export function createSpring<T extends SpringTarget>(
+  initialValue: T,
+  options: SpringOptions = {},
+): [Accessor<WidenSpringTarget<T>>, SpringSetter<WidenSpringTarget<T>>] {
+  const [signal, setSignal] = createSignal(initialValue);
+  const { stiffness = 0.15, damping = 0.8, precision = 0.01 } = options;
+
+  if (isServer) {
+    return [signal as any, ((param: any, opts: SpringSetterOptions = {}) => {
+      if (opts.hard || signal() == null || (stiffness >= 1 && damping >= 1)) {
+        setSignal(param);
+        return Promise.resolve();
+      }
+      return new Promise(() => {});
+    }) as any]
+  }
+
+  let value_current = initialValue;
+  let value_last = initialValue;
+  let value_target = initialValue;
+  let inv_mass = 1;
+  let inv_mass_recovery_rate = 0;
+  let raf_id = 0
+  let settled = true
+  let time_last = 0;
+  let time_delta = 0
+  let resolve = () => {}
+
+  const cleanup = onCleanup(() => {
+    cancelAnimationFrame(raf_id)
+    raf_id = 0
+    resolve()
+  })
+
+  const frame: FrameRequestCallback = time => {
+    time_delta = (time - time_last) * 60 / 1000
+    time_last = time
+
+    inv_mass = Math.min(inv_mass + inv_mass_recovery_rate, 1)
+    settled = true
+
+    let new_value = tick(value_last, value_current, value_target)
+    value_last = value_current
+    setSignal(value_current = new_value)
+
+    if (settled) {
+      cleanup()
+    } else {
+      raf_id = requestAnimationFrame(frame)
+    }
+  }
+
+  const set: SpringSetter<T> = (param, opts = {}) => {
+    value_target = typeof param === "function" ? param(value_current) : param;
+
+    if (opts.hard || (stiffness >= 1 && damping >= 1)) {
+      cleanup()
+      setSignal(_ => value_current = value_last = value_target);
+      return Promise.resolve();
+    }
+
+    if (opts.soft) {
+      inv_mass_recovery_rate = 1 / (typeof opts.soft === "number" ? opts.soft * 60 : 30);
+      inv_mass = 0; // Infinite mass, unaffected by spring forces.
+    }
+
+    if (raf_id === 0) {
+      time_last = performance.now()
```

Comment:
I think it should be something like `Number(document.timeline.currentTime || performance.now())` ala [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Window/requestAnimationFrame#timestamp)

In cases where `document.timeline.currentTime` is available, the first frame could have a negative `timeDelta`

Reply:
hmm youre right
I'm consistently getting negative values with `performance.now`.

![image](https://github.com/user-attachments/assets/d91a9b85-03cc-4a64-a596-7d4a36faea78)

But with `document.timeline.currentTime` it's always zero.
Which makes me think that maybe this would be better:
```ts
requestAnimationFrame(prev_time => {
	function frame(time) {
		let delta = time-prev_time
		prev_time = time
		requestAnimationFrame(frame)
	}
	requestAnimationFrame(frame)
})
```
Since the first iteration will be ran with `0` delta anyway which won't progress the animation at all, so it could be skipped.
Instead of relying on `document.timeline.currentTime`, which may be and may not be there, and falling back to `performance.now` which is wrong.